craft_parts/parts.py

```diff
@@ -320,6 +322,61 @@ def has_overlay(self) -> bool:
         """Return whether this part declares overlay content."""
         return self.spec.has_overlay
 
+    def check_partition_usage(self, partitions: List[str]) -> List[str]:
+        """Check if partitions are properly used in a part.
+
+        :param partitions: The list of valid partitions.
+
+        :returns: A list of invalid uses of partitions in the part.
+        """
+        error_list: List[str] = []
+
+        if not Features().enable_partitions:
+            return error_list
+
+        for fileset_name, fileset in [
+            ("overlay", self.spec.overlay_files),
+            # only the destination of organize filepaths use partitions
+            ("organize", self.spec.organize_files.values()),
+            ("stage", self.spec.stage_files),
+            ("prime", self.spec.prime_files),
+        ]:
+            error_list.extend(
+                self._check_partitions_in_filepaths(fileset_name, fileset, partitions)
+            )
+
+        return error_list
+
+    def _check_partitions_in_filepaths(
+        self, fileset_name: str, fileset: Iterable[str], partitions: List[str]
+    ) -> List[str]:
+        """Check if partitions are properly used in a fileset.
+
+        If a filepath begins with a parentheses, then the text inside the parentheses
+        must be a valid partition.
+
+        :param fileset_name: The name of the fileset being checked.
+        :param fileset: The list of filepaths to check.
+        :param partitions: The list of valid partitions.
+
+        :returns: A list of invalid uses of partitions in the fileset.
+        """
+        error_list = []
+
+        for filepath in fileset:
+            match = re.match("^-?\\((?P<partition>.*?)\\)", filepath)
+            if match:
+                partition = match.group("partition")
+                if partition not in partitions:
+                    error_list.append(
+                        f"    unknown partition {partition!r} in {filepath!r}"
+                    )
+
+        if error_list:
+            error_list.insert(0, f"  parts -> {self.name} -> {fileset_name}")
```

Comment:
IIRC the snapcraft.yaml documentation in the forum uses a dot as a member separator, we may want to be consistent with that in the error message.

Reply:
Reworded.  Let me know what you think.

I really would have liked to leverage the [PartSpecificationError](https://github.com/canonical/craft-parts/blob/b7c20f179afeaa8ec8cba88ce5316dce6d26a240/craft_parts/errors.py#L118), but it's designed for a single part, not a list of parts.

Reply:
Conversation continued [here](https://github.com/canonical/craft-parts/pull/493#discussion_r1263083886)